tests/cuda_custom_call/cuda_custom_call_test.py

```diff
@@ -0,0 +1,227 @@
+# This test is intentionally structured to stay close to what a standalone JAX
+# custom call integration might look like. JAX test harness is in a separate
+# section towards the end of this file. The test can be run standalone by typing
+# "make" in the directory containing this file.
+
+import ctypes
+
+import numpy as np
+
+import jax
+import jax.numpy as jnp
+from jax.extend import ffi
+from jax.lib import xla_client
+from jax._src.lib.mlir import ir
+from jaxlib.hlo_helpers import custom_call
+from jax.interpreters import mlir
+
+# start test boilerplate
+from absl.testing import absltest
+from jax._src import config
+from jax._src import test_util as jtu
+
+from absl import flags
+# escape hatch for running this example standalone.
+# note, paths to files in current working directory must be prefixed with "./".
+flags.DEFINE_string(
+    "jax_cuda_custom_call_shared_library",
+    # location where Bazel puts the runtime data dependency libfoo.so
+    "tests/cuda_custom_call/libfoo.so",
+    "Path to the custom call shared library.")
+config.parse_flags_with_absl()
+FLAGS = flags.FLAGS
+# end test boilerplate
+
+# XLA needs uppercase, "cuda" isn't recognized
+XLA_PLATFORM = "CUDA"
+
+# JAX needs lowercase, "CUDA" isn't recognized
+JAX_PLATFORM = "cuda"
+
+# 0 = original ("opaque"), 1 = FFI
+XLA_CUSTOM_CALL_API_VERSION = 1
+
+# version 4 accepts MLIR dictionaries in backend_config
+STABLEHLO_CUSTOM_CALL_API_VERSION = 4
+
+# these strings are how we identify kernels to XLA:
+# - first we register a pointer to the kernel with XLA under this name
+# - then we "tell" JAX to emit StableHLO specifying this name to XLA
+XLA_CUSTOM_CALL_TARGET_FWD = "foo-fwd"
+XLA_CUSTOM_CALL_TARGET_BWD = "foo-bwd"
+
+# independently, corresponding JAX primitives must also be named,
+# names can be different from XLA targets, here they are the same
+JAX_PRIMITIVE_FWD = "foo-fwd"
+JAX_PRIMITIVE_BWD = "foo-bwd"
+
+SHARED_LIBRARY = FLAGS.jax_cuda_custom_call_shared_library
+
+library = ctypes.cdll.LoadLibrary(SHARED_LIBRARY)
+
+#-----------------------------------------------------------------------------#
+#                              Forward pass                                   #
+#-----------------------------------------------------------------------------#
```

Comment:
We don't typically use banner comments like this in JAX. I think these should be removed.

Reply:
An end-to-end example like this can be overwhelming to read, I added the banner comments to make it more "skimmable" :)

Reply:
(I don't typically use banner comments either, but since this file is kind of dense and is not part of "core" JAX code, I think it's an improvement to have these visual separators)